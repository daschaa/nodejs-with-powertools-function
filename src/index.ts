export * from './nodejs-with-powertools-function';
