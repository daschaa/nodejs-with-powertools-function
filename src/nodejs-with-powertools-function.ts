import { Stack } from "aws-cdk-lib";
import { LayerVersion } from "aws-cdk-lib/aws-lambda";
import {
  NodejsFunction,
  type NodejsFunctionProps,
} from "aws-cdk-lib/aws-lambda-nodejs";
import { Construct } from "constructs";

/**
 * The supported environment variables for Powertools
 */
export type PowertoolsEnvironmentVariables =
  | "POWERTOOLS_SERVICE_NAME"
  | "POWERTOOLS_METRICS_NAMESPACE"
  | "POWERTOOLS_TRACE_ENABLED"
  | "POWERTOOLS_TRACER_CAPTURE_RESPONSE"
  | "POWERTOOLS_TRACER_CAPTURE_ERROR"
  | "POWERTOOLS_TRACER_CAPTURE_HTTPS_REQUESTS"
  | "POWERTOOLS_LOGGER_LOG_EVENT"
  | "POWERTOOLS_LOGGER_SAMPLE_RATE"
  | "POWERTOOLS_DEV"
  | "POWERTOOLS_LOG_LEVEL"
  | "POWERTOOLS_PARAMETERS_MAX_AGE"
  | "POWERTOOLS_PARAMETERS_SSM_DECRYPT"
  | "POWERTOOLS_IDEMPOTENCY_DISABLED";

export type NodejsWithPowertoolsFunctionProps = {
  /**
   * Sets the environment variables that are specific for Lambda Powertools.
   *
   * @see https://docs.powertools.aws.dev/lambda/typescript/latest/#environment-variables
   */
  readonly powertoolsEnvironmentVariables?: {
    [key in PowertoolsEnvironmentVariables]?: string;
  };
  /**
   * Sets the version of the Powertools layer in the layer version ARN.
   *
   * @default 10
   */
  readonly powertoolsVersion?: string;
};

/**
 * Create a NodejsFunction that has the Lambda Powertools Layer enabled.
 */
export class NodejsWithPowertoolsFunction extends Construct {

  /**
   * The underlying NodejsFunction construct.
   */
  public readonly fn: NodejsFunction;

  constructor(
    scope: Construct,
    id: string,
    props?: NodejsFunctionProps & NodejsWithPowertoolsFunctionProps,
  ) {
    super(scope, id);
    const powertoolsLayer = LayerVersion.fromLayerVersionArn(
      this,
      "PowertoolsLayer",
      `arn:aws:lambda:${Stack.of(this).region}:094274105915:layer:AWSLambdaPowertoolsTypeScriptV2:${props?.powertoolsVersion ?? "10"}`,
    );

    this.fn = new NodejsFunction(this, "Resource", {
      ...props,
      layers: [powertoolsLayer],
      environment: {
        ...props?.environment,
        ...props?.powertoolsEnvironmentVariables,
      },
      bundling: {
        ...props?.bundling,
        externalModules: [
          ...(props?.bundling?.externalModules ?? []),
          "@aws-lambda-powertools/*",
          "@aws-sdk/*",
        ],
      },
    });
  }
}
