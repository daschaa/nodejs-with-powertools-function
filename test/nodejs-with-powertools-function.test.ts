import { NodejsWithPowertoolsFunction } from '../src/nodejs-with-powertools-function';
import { Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

test('render initial function', () => {
  const stack = new Stack();
  new NodejsWithPowertoolsFunction(stack, 'TestFunction', {
    entry: __dirname + '/test-function.ts',
    powertoolsEnvironmentVariables: {
      POWERTOOLS_SERVICE_NAME: 'my-cool-service'
    }
  });
  expect(Template.fromStack(stack).toJSON()).toMatchSnapshot();
});
