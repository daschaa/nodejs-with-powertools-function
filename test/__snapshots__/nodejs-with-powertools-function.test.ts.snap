// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render initial function 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "TestFunction4F6A4A33": {
      "DependsOn": [
        "TestFunctionServiceRole937A341F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "aba5618200a920b21b715c214454cf0e0024ef0aa3e2ef50a7ed41a15664c8d6.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "POWERTOOLS_SERVICE_NAME": "my-cool-service",
          },
        },
        "Handler": "index.handler",
        "Layers": [
          {
            "Fn::Join": [
              "",
              [
                "arn:aws:lambda:",
                {
                  "Ref": "AWS::Region",
                },
                ":094274105915:layer:AWSLambdaPowertoolsTypeScriptV2:10",
              ],
            ],
          },
        ],
        "Role": {
          "Fn::GetAtt": [
            "TestFunctionServiceRole937A341F",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestFunctionServiceRole937A341F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
